abstract type of the document

matklad opened the pull request:
Попробовал абстрагировать тип документа(до этого он был `type Document = HashMap Label Value`).

Хочется иметь возможность заменить представление в виде хэштаблицы на что-нибудь другое, чтобы при этом не сломалось ничего ни внутри, ни снаружи.

Есть предположение, что, так как многие JSON-документы соедржат не очень много ключей, можно придумать что-то более быстрое, чем HashTable.

Не очень понятно, какие функции про Document надо экспортировать(например мне пришлось вытащить `Document.fromList` для тестов, но это, очевидно, плохо).

matklad committed:
abstract type of the document

knsd commented:
Думаю, что документ можно не параметризовывать. А почему `fromList` плохо?

superbobry commented on src/Data/Bson/Binary.hs:
```diff
@@ -144,7 +145,7 @@ getDocument :: Get Document
     l <- fromIntegral <$> getWord32le
     bytes <- getLazyByteString $ l - 4
-    return $ HashMap.fromList $ runGet getFields bytes
+    return $ Doc.fromList $ runGet getFields bytes
```
Было бы клево переписать это так, чтобы fromList ушел. Тобишь через что-то foldM подобное.

superbobry commented on src/Data/Bson/Document.hs:
```diff
@@ -0,0 +1,35 @@
+lookup k (Document doc) = DocImpl.lookup k doc
+
+lookupDefault :: (Eq k, Hashable k) => v -> k -> Document k v -> v
+lookupDefault v k (Document doc)= DocImpl.lookupDefault v k doc
```
`DocImpl`, scalaish bastard!

matklad commented:
Да, документ надо не параметризовать, потому что, зная тип ключа, можно будет какой-нибудь более интересный способ  хранения реализовать. 

У меня документ с параметрами, потому что я не придумал, как избежать циклических импортов: Если `Document` будет знать про `Label` и `Value`, то надо в `Document` импортировать `Types`, но `Types` сами импортируют `Document` =( Как сделать хорошо? `Document` и `Value` в любом случае взаимно-рекурсивные, но хочется, чтобы Document был отдельным =)

Как вариант, можно сделать Document полиморфным только по `Value`.

matklad commented:
> почему fromList плохо?

Потому что, если экспортируется fromList, то почему бы и не экспортнуть все остальные функции из Document?

Вопрос на самом деле в том, какой у Document должен быть публичный интерфейс. Если Document не прозрачный, то надо убрать fromList и что-нибудь с тестами придумать. Если у Document есть интерфейс мапки, то надо этот интерфес сделать, со всеми fold и get.

knsd committed:
Some mutually recursive modules

knsd commented:
Нагородил взаимно-рекурсивных модулей, чтобы не параметризовывать `Document`.

matklad commented:
Смотри, `Label` надо будет ещё протаскивать в модуль-имплементацию Document, иначе это никакого профита не даст. Большой огород получится.

knsd committed:
Remove parametrisation from Document in implementation module

knsd commented:
Слушай, ну вроде просто протащить, достаточно импортировать с `{-# SOURCE #-}` прагмой. бтв ты планируешь несколько модулей имплементации? А то можно имплементацию прямо в `Data.Bson.Document` и писать.

matklad commented:
> Слушай, ну вроде просто протащить, достаточно импортировать с {-# SOURCE #-} прагмой.

Ок. Хотя пока идея с параметризованным Value и импортом одного Label мне всё же чуть больше нравится.

> ты планируешь несколько модулей имплементации?

В итоге должен быть один, самый лучший, но его ещё надо найти. Мысль писать имплементацию 
прямо в Document мне нравится. Ну и ничего не мешает там написать две имплементации, если захочется =)

matklad commented:
какой у Document должен быть публичный интерфейс?

knsd committed:
Move document implementation in Data.Bson.Document

knsd commented:
Перенёс имплементацию.
Мне кажется, да, должен быть интерфейс как в хешмепе, по крайней мере в случае аесона это было удобно.

knsd committed:
Remove unused import

knsd committed:
Remove unused import

matklad committed:
use `document` instead of `fromList` in tests.

This allows to remove `fromList` from public API.
May be use `document` in Binary.hs as well?

matklad commented:
> по крайней мере в случае аесона это было удобно.

Насколько я понимаю, у аесона type Object = HashMap --- как раз то, чего хотелось бы избежать, чтобы иметь возможность поменять имплементацию =)

superbobry commented:
newtype!

knsd commented:
Хм, сори, не думал, что Гитхаб закроет при мердже в мастер.
